debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/infinispan/InfinispanLogMinerTransactionCache.java

```diff
@@ -0,0 +1,247 @@
+        waitForAllWrites();
+
+        eventIdsByTransactionId.get(transaction.getTransactionId())
+                .descendingSet()
+                .stream()
+                .map(transaction::getEventId)
+                .forEach(eventCache::remove);
+
+        eventIdsByTransactionId.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean removeTransactionEventWithRowId(InfinispanTransaction transaction, String rowId) {
+        waitForAllWrites();
+
+        final TreeSet<Integer> eventIds = eventIdsByTransactionId.get(transaction.getTransactionId());
+        return eventIds.descendingSet().stream()
+                .map(i -> Map.entry(i, transaction.getEventId(i)))
+                .filter(entry -> {
+                    final LogMinerEvent event = eventCache.get(entry.getValue());
+                    return event != null && event.getRowId().equals(rowId);
+                })
+                .findFirst()
+                .map(entry -> {
+                    eventCache.remove(entry.getValue());
+                    eventIds.remove(entry.getKey());
+                    return true;
+                })
+                .orElse(false);
+    }
+
+    @Override
+    public boolean containsTransactionEvent(InfinispanTransaction transaction, int eventKey) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId()).contains(eventKey);
+    }
+
+    @Override
+    public int getTransactionEventCount(InfinispanTransaction transaction) {
+        return eventIdsByTransactionId.get(transaction.getTransactionId()).size();
+    }
+
+    @Override
+    public int getTransactionEvents() {
+        return eventIdsByTransactionId.values().stream().mapToInt(Set::size).sum();
+    }
+
+    @Override
+    public void clear() {
+        // Wait for all writes before we clear the cache.
+        waitForAllWrites();
+
+        transactionCache.clear();
+        eventCache.clear();
+        eventIdsByTransactionId.clear();
+    }
+
+    @Override
+    public void resetTransactionToStart(InfinispanTransaction transaction) {
+        super.resetTransactionToStart(transaction);
+        syncTransaction(transaction);
+    }
+
+    @Override
+    public void syncTransaction(InfinispanTransaction transaction) {
+        // todo:
+        // Perhaps we can look at pulling number of events out of Transaction and let that
+        // be managed in the cache's heap, in which case we can avoid this put.
+
+        // Necessary to synchronize state
+        putAsync(transactionCache, transaction.getTransactionId(), transaction);
+    }
+
+    @Override
+    public void close() throws Exception {
+        // Makes sure during shutdown that all pending writes are flushed
+        waitForAllWrites();
+    }
+
+    private void primeEventCountsByTransactionId() {
+        // Primes the heap-based cache if the Infinispan disk caches contained data on start-up
+        transactions(stream -> stream.forEach(transaction -> {
```

Comment:
Can we avoid streaming over the transactions by simply streaming overall events and then looking up the transaction for each event?  Unsure of performance implications here, but I'm thinking it might be better.

Reply:
@Naros - but I assume this is at startup, so may not be a major concern.

Reply:
This is performed at startup. As for performance, you're probably right. Worse case, there would be 1 transaction lookup for each event effectively evicting each transaction at some point, but likely have far more events associated with a single transaction in practice :+1: